Add trailing window aggregator

cdg-stripe opened the pull request:
#### Description

Add WindowMonoid utility for aggregating over finite window. Window case class stores elements with a queue. The monoid combines windows while trimming their elements to the appropriate size. The monoid is also geared to aggregate elements with the group property more efficiently, as we can subtract from the total elements we push out of the queue upon aggregation.

#### Testing

Ran tests locally. Unsure how tests are incorporated here.

r? @johnynek

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/twitter/algebird?pullRequest=649) <br/>All committers have signed the CLA.

johnynek commented on algebird-core/src/main/scala/com/twitter/algebird/Window.scala:
```diff
@@ -0,0 +1,118 @@
+case class WindowMonoid[T](
+    windowSize: Int
+)(
+    implicit m: Monoid[T],
```
Let’s not use this m. You can get the zero from the p by case matching.

MansurAshraf commented on algebird-core/src/main/scala/com/twitter/algebird/Window.scala:
```diff
@@ -0,0 +1,118 @@
+ * @param q (optional) queue to override default q containing input "total"
+ */
+
+case class Window[T](total: T, q: Queue[T] = Queue()) {
```
This looks great! What do you think about allowing user to pass a `PriorityQueue` instead of just a Queue (AFAIK scala PQ doesnt extend Queue). The way code is written right now, its implying that I need a commutative monoid to guarantee deterministic behavior or I need to control the order in  which items are inserted in the window. When I work with windows they usually have  some sort of ordering defined (last 3 distinct IPs used by a user in last one hour for example), It would be nice to allow users to  preserve the ordering somehow IMO

johnynek replied:
actually, I think `Window[T](total: T, items: Queue[T])` is better. We don't really want people to do something like `Window(100)` do we? that will have no items in the queue, and I imagine break the laws.

cdg-stripe replied:
If you look at `items`, it fills in the queue if none was provided, i.e meant for initialization. The intent is to have a simple interface where you can wrap a window around a single object and sum in a sensible way:
```
implicit def wm[T] = WindowMonoid[T](2)

Window(1) + Window(2) + Window(3) == Window(5)
```
Needing to specify the queue would make this redundant 
```
Window(1, Queue(1)) + ...
```

Tbh I'm not a big fan of even having to expose the queue, since you could change the implementation in the future (e.g. a PriorityQueue). A more general approach might use a `PushPopper` interface, but that's probably overkill.

MansurAshraf commented on algebird-core/src/main/scala/com/twitter/algebird/Window.scala:
```diff
@@ -0,0 +1,118 @@
+    if (b.items.size >= windowSize) {
+      var total: T = b.total
+      var q = b.items
+      while (q.size > windowSize) {
```
Not sure if its possible to do it in this implementation but it would be nice to generalize how elements are evicted from the window. Right now it seems like I can only do fixed windows but I would really like to do time based rolling windows as well. For example: Rolling window of 1 hour and we evict element based on time instead of size

cdg-stripe replied:
My thought was to eventually have two classes/monoids, based on whether we want aggregations over fixed items or fixed time.

```
// Monoid defines aggregation for fixed window length, e.g. last 28 elements. Queue fixed to 28.
case class Window[T](total:T, q: Queue[T])

// Monoid defines aggregation for fixed time, e.g. last 28 days. Queue can grow arbitrarily.
case class WindowWithTs(lastestTime: Double, total: T, q: Queue[(Double, T)])
```

Does that seem sensible?

MansurAshraf replied:
That seems reasonable but I am still not sure why we cant use `ProrityQueue` instead of `Queue`? As a user I would like to control the ordering and if I am passing you a PQ, I can do that by specifying my own comparator

cdg-stripe commented:
@MansurAshraf: I'm trying to grasp whether we need a fully general implementation up front, where the user inputs their own queue/priority. As mentioned above my thoughts are to have trailing windows based on one of:

- Fixed number of elements
- Fixed time interval (not implemented here, but can in a separate PR)

The "last N ip-addresses" problem is interesting. But rather than making this implementation more flexible/abstract, it might make sense to have a `LastNDistinct` monoid. I suspect this approach will be easier to test and implement.

cdg-stripe commented:
@johnynek:

Added tests and removed the redundant `implicit m: Monoid[T]`. I also tried out making `Window` a `class` instead of a `case class`. 
```
class Window[T](t: T, q: Queue[T] = Queue()) {
  def total = t
  def items = if (q.size > 0) q else Queue(total)
  def size = items.size
}
```
This makes it clear that q is optional and never accessible after instantiation. If no q is given, items default `Queue(total)`, the expected behavior when instantiating a window from a single item.

johnynek commented on algebird-core/src/main/scala/com/twitter/algebird/Window.scala:
```diff
@@ -0,0 +1,120 @@
+}
+
+/*
+  Example usage:
```
can you make this a scaladoc comment on the case class?

johnynek commented on algebird-test/src/test/scala/com/twitter/algebird/WindowLawsTest.scala:
```diff
@@ -0,0 +1,20 @@
+      val ts = ts0.takeRight(n)
+      val mon = new WindowMonoid(n)
+      assert(mon.sum(ts0.map( Window(_) )).total == ts.sum)
+    }
```
can we have a second one:
```scala
forAll { (ts0: List[Int], ts1: List[Int], n: Int) =>
  val expected = Queue((ts0 ::: ts1).takeRight(n): _*)
  val mon = new WindowMonoid(n)
  val got = mon.plus(Window(ts0.sum, Queue(ts0: _*)), Window(ts1.sum, Queue(ts1:_*)))
  assert(got == expected)
}
```

cdg-stripe replied:
Done

johnynek commented on algebird-core/src/main/scala/com/twitter/algebird/Window.scala:
```diff
@@ -0,0 +1,120 @@
+/**
+ * Provides a natural monoid for combining windows truncated to some window size.
+ *
+ * @param windowSize
```
can you add more comment to this javaDoc param?

johnynek commented on algebird-core/src/main/scala/com/twitter/algebird/Window.scala:
```diff
@@ -0,0 +1,120 @@
+      case Priority.Preferred(g) => Window(g.zero)
+      case Priority.Fallback(m)  => Window(m.zero)
+    }
+
```
what about a
```scala
  def fromIterable[T](ts: Iterable[T]): Window[T] = {
    val monT: Monoid[T] = p.join
    val right = ts.toList.takeRight(windowSize)
    val total = monT.sum(right)
    Window(total, Queue.empty[T] ++ right)
  }
```

cdg-stripe replied:
@johnynek: I did this, and generalized to `Traversable`.

johnynek commented on algebird-core/src/main/scala/com/twitter/algebird/Window.scala:
```diff
@@ -0,0 +1,120 @@
+      case Priority.Preferred(g) => plusG(a, b)(g)
+      case Priority.Fallback(m)  => plusM(a, b)(m)
+    }
+
```
I think you can optimize `sumOption` somewhat here: you can take blocks of data and ignore the left if the right is full. PS: scalding uses this if it exists, so overridding sumOption on Semigroup (or any subclass) can be a very nice performance win on map/reduce.

```scala
def sumOption(ws: TraversableOnce[Window[T]]): Option[Window[T]] =
  if (ws.isEmpty) None
  else {
    val it = ws.toIterator
    var size = 0
    var queue = Queue.empty[Window[T]]
    while (it.hasNext) {
      val n = it.next
      queue = queue :+ n
      size += n.size
      val tailSize = size - queue.head.size
      if (tailSize >= windowSize) {
        queue = queue.tail
        size = tailSize
      }
    }
    // now we only have to merge the queue:
    queue.tail.foldLeft(queue.head)(plus(_, _))
  }
```
I think this can be a pretty big win.

cdg-stripe replied:
@johnynek: Overwrote sumOption below.

johnynek commented:
this PR: https://github.com/twitter/algebird/pull/650 seems to fix the ruby issue on travis.

johnynek commented:
(nothing like non-hermetic builds to make you feel real good, all day long).

johnynek commented on algebird-core/src/main/scala/com/twitter/algebird/Window.scala:
```diff
@@ -0,0 +1,145 @@
+    if(ws.isEmpty) None
+    else {
+      val it = ws.toIterator
+      var queue = Queue[T]()
```
.empty is more efficient since you avoid allocating the varargs wrapper.

johnynek commented on algebird-core/src/main/scala/com/twitter/algebird/Window.scala:
```diff
@@ -0,0 +1,145 @@
+      while (it.hasNext) {
+        queue = (queue ++ it.next.items).takeRight(windowSize)
+      }
+      Some(fromTraversable(queue))
```
Using fromTraversable here is ignoring that you already have a queue. I think summing the queue and reusing it will be a significant performance improvement

johnynek commented on algebird-core/src/main/scala/com/twitter/algebird/Window.scala:
```diff
@@ -0,0 +1,145 @@
+
+object Window {
+  def apply[T](v: T): Window[T] = Window[T](v, Queue[T](v))
+  def from[T](ts: Traversable[T])(implicit m: WindowMonoid[T]) = m.fromTraversable(ts)
```
Can we use Iterable? Scala 2.13 is changing the collections to avoid Traversable: https://www.scala-lang.org/blog/2017/02/28/collections-rework.html#traversable-and-iterable

cdg-stripe commented:
@johnynek:

Added more tests. Caught that the zero element should have an empty queue, where originally it was `Queue[T](T.zero)`. Also made the monoid on group elements more efficient. I dug around to see if there's anything else worth tweaking, but I think it's in a decent spot.

Also what are you thoughts about Scala 2.13, when we can have a signature `Window[Size, T]`? Might be okay have `Window[T]` and `WindowN[Size, T]` case classes.

johnynek commented:
yeah, I think in scala 2.13 we will have to revisit a lot of stuff, but the compatibility story is open since we will need to support 2.12 for a while, I guess (we aren't even on 2.12 yet at Stripe).

johnynek commented on algebird-core/src/main/scala/com/twitter/algebird/Window.scala:
```diff
@@ -0,0 +1,144 @@
+
+  require(windowSize >= 1, "Windows must have positive sizes")
+
+  def zero = p.fold(g => Window[T](g.zero, Queue.empty[T]))(m => Window(m.zero, Queue.empty[T]))
```
can we make this a `val` so we don't have to do the computation and allocation again on each call?

johnynek commented on algebird-core/src/main/scala/com/twitter/algebird/Window.scala:
```diff
@@ -0,0 +1,144 @@
+  def zero = p.fold(g => Window[T](g.zero, Queue.empty[T]))(m => Window(m.zero, Queue.empty[T]))
+
+  def plus(a: Window[T], b: Window[T]): Window[T] =
+    p.fold(g => plusG(a, b)(g))(m => plusM(a, b)(m))
```
for performance (not calling fold on each plus call) we are probably better to have two subclasses of `WindowMonoid` which could be an abstract class. We can decide once when we allocate which path to choose. Then the jit should be able to inline both instances. Sadly, the jvm can't optimize the above very well.

johnynek commented:
published in 0.13.4:

https://github.com/twitter/algebird/releases/tag/v0.13.4

johnynek commented:
@MansurAshraf note if you want the most recent K things, TopKMonoid should work:

https://github.com/twitter/algebird/blob/develop/algebird-core/src/main/scala/com/twitter/algebird/TopKMonoid.scala

(so would priorityqueue I guess).
